src/rpp/rpp/schedulers/trampoline_scheduler.hpp

```diff
@@ -0,0 +1,102 @@
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/schedulers/fwd.hpp>                       // own forwarding
+#include <rpp/schedulers/details/worker.hpp>            // worker
+#include <rpp/subscriptions/composite_subscription.hpp> // lifetime
+#include <rpp/subscriptions/subscription_guard.hpp>     // lifetime
+#include <rpp/schedulers/details/queue_worker_state.hpp>// state
+
+#include <concepts>
+#include <chrono>
+#include <functional>
+#include <thread>
+
+namespace rpp::schedulers
+{
+
+/**
+ * \brief scheduler which schedules execution of schedulables via queueing tasks to the caller thread with priority to time_point and order
+ * \warning Creates new thread for each "create_worker" call. Any scheduled task will be queued to created thread for execution with respect to time_point and number of task
+ * \ingroup schedulers
+ */
+class trampoline final : public details::scheduler_tag
+{
+public:
+    class worker_strategy_state : public std::enable_shared_from_this<worker_strategy_state>
+    {
+    public:
+        explicit worker_strategy_state(const rpp::composite_subscription& subscription) : m_sub(subscription) {
+        }
+
+        // Not copyable nor movable
+        worker_strategy_state(const worker_strategy_state&) = delete;
+        worker_strategy_state(worker_strategy_state&&) noexcept = delete;
+
+        void defer_at(time_point time_point, std::invocable auto&& fn)
+        {
+            thread_local bool queue_in_use = false;
+            thread_local details::queue_worker_state queue{};
+
+            if (!m_sub->is_subscribed())
+                return;
+
+            queue.emplace(time_point, std::forward<decltype(fn)>(fn));
+            if (queue_in_use)
+                return;
+
+            try {
+                queue_in_use = true;
+                while (!queue.is_empty() && m_sub->is_subscribed())
+                {
+                    std::function<void()> schedulable{};
+                    if (queue.pop_if_ready(schedulable))
+                    {
+                        schedulable();
+                        schedulable = {};
+                    }
+                }
+            } catch (const std::exception&) {
+                // no-op
+            }
+            queue_in_use = false;
+        }
+
+    private:
+        rpp::subscription_guard m_sub = rpp::subscription_base::empty();
+    };
+
+    class worker_strategy
+    {
+    public:
+        explicit worker_strategy(
+                const rpp::composite_subscription& subscription)
+            : m_state{std::make_shared<worker_strategy_state>(subscription)} {};
+
+        void defer_at(time_point time_point, std::invocable auto&& fn) const
+        {
+            m_state->defer_at(time_point, std::forward<decltype(fn)>(fn));
+        }
+
+        static time_point now() { return clock_type::now(); }
+
+    private:
+        std::shared_ptr<worker_strategy_state> m_state{};
```

Comment:
Can't see if we really need shared_ptr here. composite_subscription is shared_ptr itself

Reply:
I see. I'll reduce one layer of classes.